### PR TITLE
Library tutorial

### PR DESCRIPTION
**Pull request by a-stepin**
Add functionality to add, delete, show certain amount of comments
Add page that displays map
Add map that displays big foot sighting locations

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/data/CommentBlock.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  /** List of descriptions of comments, e.g. "Anonymous said: [new line] 'Hello World' " */
+  private final List<String> history = new ArrayList<>();
+
+  public List<String> getHistory(){ return history; }
+
+  public void logComment(String name, String comment) {
+    history.add(name + " said: \n" + "'" + comment + "'");
+  }
+
+  private class Comment{
```
Comment (Felomeng) [inline portfolio/src/main/java/com/google/sps/data/CommentBlock.java:35]: why have this class here?

File: portfolio/src/main/java/com/google/sps/data/CommentBlock.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ */
+public class CommentBlock {
+
+  /** List of descriptions of comments, e.g. "Anonymous said: [new line] 'Hello World' " */
+  private final List<String> history = new ArrayList<>();
+
+  public List<String> getHistory(){ return history; }
+
+  public void logComment(String name, String comment) {
+    history.add(name + " said: \n" + "'" + comment + "'");
```
Comment (Felomeng) [inline portfolio/src/main/java/com/google/sps/data/CommentBlock.java:32]: String.format or StringBuilder would have much better performance.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteServlet.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
```
Comment (Felomeng) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteServlet.java:25]: clean up duplicate imports

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -14,48 +14,112 @@
...
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
```
Comment (Felomeng) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:25]: duplicate imports

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -14,48 +14,112 @@
...
+            numComments = Integer.parseInt(numComString);
+
+        } catch (NumberFormatException e) {
+            numComments = -1;
+        }
+        int numComLogged = 0;
+
+        for (Entity entity : results.asIterable()) {
+            if(numComments == 0 || numComments > numComLogged){
+                String comment = (String) entity.getProperty("comment");
```
Comment (Felomeng) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:64]: can comment or name be null? what will happen?

It is not a good idea to convert this way. Maybe String.valueOf

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -14,48 +14,112 @@
...
+    }
+
+
+    /**
+    * Converts a List of Strings into a JSON string using the Gson library. Note: We first added
+    * the Gson library dependency to pom.xml.
+    */
+    private String convertToJson(List<String> messages) {
+        Gson gson = new Gson();
+        String json = gson.toJson(messages);
```
Comment (Felomeng) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:118]: feel like new Gson().toJson(msgs); is good enough?
And were we to introduce an util function, make it static and in a separate class

Comment (a-stepin) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:118]: My justification for this is that this method is copy-and-pasted from the example code that we are provided in the tutorial. However, if need be, I can change it.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -14,48 +14,112 @@
...
+        try {
+            numComments = Integer.parseInt(numComString);
+
+        } catch (NumberFormatException e) {
+            numComments = -1;
+        }
+        int numComLogged = 0;
+
+        for (Entity entity : results.asIterable()) {
+            if(numComments == 0 || numComments > numComLogged){
```
Comment (salehhamadeh) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:59]: Did you mean `numComments == -1`?

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -31,33 +31,91 @@ function addRandomGreeting() {
...
-        createListElement('First Message: ' + stats.firstMessage));
-    statsListElement.appendChild(
-        createListElement('Second Message: ' + stats.secondMessage));
-    statsListElement.appendChild(
-        createListElement('Third Message: ' + stats.thirdMessage));
-  });
+function getComments() {
+
+    let comVal = document.getElementById('numComments').value;
+    console.log(comVal)
```
Comment (salehhamadeh) [inline portfolio/src/main/webapp/script.js:46]: nit: Remove console.log.

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -31,33 +31,91 @@ function addRandomGreeting() {
...
+    fetch(`/data?numComments=${comVal}`).then(response => response.json()).then((comments) => {
+        const historyEl = document.getElementById('past-comments');
+        comments.forEach((line) => {
+            historyEl.appendChild(createListElement(line));
+        });
+    }); 
+}
+
+function setValComments(e){
+    console.log(e.sender.value)
```
Comment (salehhamadeh) [inline portfolio/src/main/webapp/script.js:56]: Is this used? If not please remove from this PR.

**Pull request merged**